packages/flutter_tools/lib/src/flutter_plugins.dart

```diff
@@ -1137,127 +1137,176 @@ List<PluginInterfaceResolution> resolvePlatformImplementation(
+    if (defaultImplPluginName == null && !hasInlineDartImplementation) {
+      // Skip native inline PluginPlatform implementation.
+      return (null, null);
+    }
+    if (defaultImplPluginName != null) {
+      return (null, defaultImplPluginName);
+    } else {
+      // An app-facing package (i.e., one with no 'implements') with an
+      // inline implementation should be its own default implementation.
+      // Desktop platforms originally did not work that way, and enabling
+      // it unconditionally would break existing published plugins, so
+      // only treat it as such if either:
+      // - the platform is not desktop, or
+      // - the plugin requires at least Flutter 2.11 (when this opt-in logic
+      //   was added), so that existing plugins continue to work.
+      // See https://github.com/flutter/flutter/issues/87862 for details.
+      final bool isDesktop = platformKey == 'linux' || platformKey == 'macos' || platformKey == 'windows';
+      final semver.VersionConstraint? flutterConstraint = plugin.flutterConstraint;
+      final semver.Version? minFlutterVersion = flutterConstraint != null &&
+          flutterConstraint is semver.VersionRange ? flutterConstraint.min : null;
+      final bool hasMinVersionForImplementsRequirement = minFlutterVersion != null &&
+          minFlutterVersion.compareTo(semver.Version(2, 11, 0)) >= 0;
+      if (!isDesktop || hasMinVersionForImplementsRequirement) {
+        // Dart plugin serves as its own default implementation.
+        implementsPackage = plugin.name;
+        defaultImplPluginName = plugin.name;
+        return (implementsPackage, defaultImplPluginName);
+      } else {
+        // If the inline dart implementation is desktop and it's flutter min
+        // constraint isn't fulfilled, it isn't eligible for auto-registration.
+        return (null, null);
+      }
+    }
+  }
+
+  if (!hasInlineDartImplementation) {
+    // If there's no Dart implementation, there's nothing to register.
+    return (null, null);
+  }
+
+  // If it hasn't been skipped, it's a candidate for auto-registration,
+  // means it can serve as implementation for [implementsPackage],
+  // so add it as implementation candidate.
+  return (implementsPackage, null);
+}
+
+/// Get the resolved plugin name serving as implementation for a plugin.
+Plugin? _resolveImplementationOfPlugin(
+  String platformKey,
+  MapEntry<String, List<Plugin>> implCandidatesEntry, {
+  String? defaultPackageName,
+}) {
+  final List<Plugin> candidates = implCandidatesEntry.value;
+  // If there's only one candidate, use it.
+  if (candidates.length == 1) {
+    return candidates.first;
+  }
+  // Next, try direct dependencies of the resolving application.
+  final Iterable<Plugin> directDependencies = candidates.where((Plugin plugin) {
+    return plugin.isDirectDependency;
+  });
+  if (directDependencies.isNotEmpty) {
+    if (directDependencies.length > 1) {
```

Comment:
Is an app-facing plugin with an inline implementation included in the candidates list, or only the default list? If the former, it seems like this won't handle the case where an app wants to override an implementation that is inline in the app-facing package, since the app-facing package will be a direct dependency.

Reply:
I created such case and it is indeed failing (for both: current `main` and this branch):


<details>
  <summary>dart_plugin_test.dart</summary>
  

```dart
      testWithoutContext('selects user selected implementation despite inline implementation on mobile', () async {
        final Set<String> directDependencies = <String>{
          'user_selected_url_launcher_implementation',
          'url_launcher',
        };

        final List<PluginInterfaceResolution> resolutions = resolvePlatformImplementation(<Plugin>[
          Plugin.fromYaml(
            'url_launcher',
            '',
            YamlMap.wrap(<String, dynamic>{
              'platforms': <String, dynamic>{
                'android': <String, dynamic>{
                  'dartPluginClass': 'UrlLauncherAndroid',
                },
                'ios': <String, dynamic>{
                  'dartPluginClass': 'UrlLauncherIos',
                },
              },
            }),
            null,
            <String>[],
            fileSystem: fs,
            appDependencies: directDependencies,
          ),
          Plugin.fromYaml(
            'user_selected_url_launcher_implementation',
            '',
            YamlMap.wrap(<String, dynamic>{
              'implements': 'url_launcher',
              'platforms': <String, dynamic>{
                'android': <String, dynamic>{
                  'dartPluginClass': 'UrlLauncherAndroid',
                },
              },
            }),
            null,
            <String>[],
            fileSystem: fs,
            appDependencies: directDependencies,
          ),
        ]);
        expect(resolutions.length, equals(2));
        expect(resolutions[0].toMap(), equals(
            <String, String>{
              'pluginName': 'user_selected_url_launcher_implementation',
              'dartClass': 'UrlLauncherAndroid',
              'platform': 'android',
            })
        );
        expect(resolutions[1].toMap(), equals(
            <String, String>{
              'pluginName': 'url_launcher',
              'dartClass': 'UrlLauncherIos',
              'platform': 'ios',
            })
        );
      });
```

</details>

I think in the [design doc](https://docs.google.com/document/d/1LD7QjmzJZLCopUrFAAE98wOUQpjmguyGTN2wd_89Srs/edit#), this case was not explicitly considered (or I missed something).

I would rather fix this in its own PR(?)